File: frontend/src/components/benchmark/content/bellman_ford/implementations/bellman_ford.cpp
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <climits>
#include <optional>

using namespace std;

class BellmanFord {
private:
    struct Edge {
        int u, v, weight;
    };
    
    vector<Edge> edges;
    unordered_set<int> vertices;

public:
    void add_edge(int u, int v, int weight) {
        edges.push_back({u, v, weight});
        vertices.insert(u);
        vertices.insert(v);
    }
    
    void build_from_edges(const vector<tuple<int, int, int>>& edge_list) {
        for (const auto& [u, v, w] : edge_list) {
            add_edge(u, v, w);
        }
    }
    
    optional<unordered_map<int, int>> shortest_path(int source) {
        if (vertices.find(source) == vertices.end()) {
            return nullopt;
        }
        
        unordered_map<int, int> dist;
        for (int v : vertices) {
            dist[v] = INT_MAX;
        }
        dist[source] = 0;
        
        for (size_t i = 0; i < vertices.size() - 1; i++) {
            bool updated = false;
            for (const auto& edge : edges) {
                if (dist[edge.u] != INT_MAX && dist[edge.u] + edge.weight < dist[edge.v]) {
                    dist[edge.v] = dist[edge.u] + edge.weight;
                    updated = true;
                }
            }
            if (!updated) break;
        }
        
        for (const auto& edge : edges) {
            if (dist[edge.u] != INT_MAX && dist[edge.u] + edge.weight < dist[edge.v]) {
                return nullopt;
            }
        }
        
        return dist;
    }
    
    bool has_negative_cycle() {
        int start = *vertices.begin();
        return !shortest_path(start).has_value();
    }
};

File: frontend/src/components/benchmark/content/bellman_ford/implementations/bellman_ford.go
package main

import (
	"math"
)

type Edge struct {
	U, V, Weight int
}

type BellmanFord struct {
	edges    []Edge
	vertices map[int]bool
}

func NewBellmanFord() *BellmanFord {
	return &BellmanFord{
		edges:    []Edge{},
		vertices: make(map[int]bool),
	}
}

func (bf *BellmanFord) AddEdge(u, v, weight int) {
	bf.edges = append(bf.edges, Edge{U: u, V: v, Weight: weight})
	bf.vertices[u] = true
	bf.vertices[v] = true
}

func (bf *BellmanFord) BuildFromEdges(edgeList [][3]int) {
	for _, edge := range edgeList {
		bf.AddEdge(edge[0], edge[1], edge[2])
	}
}

func (bf *BellmanFord) ShortestPath(source int) map[int]int {
	if !bf.vertices[source] {
		return nil
	}

	dist := make(map[int]int)
	for v := range bf.vertices {
		dist[v] = math.MaxInt32
	}
	dist[source] = 0

	for i := 0; i < len(bf.vertices)-1; i++ {
		updated := false
		for _, edge := range bf.edges {
			if dist[edge.U] != math.MaxInt32 && dist[edge.U]+edge.Weight < dist[edge.V] {
				dist[edge.V] = dist[edge.U] + edge.Weight
				updated = true
			}
		}
		if !updated {
			break
		}
	}

	for _, edge := range bf.edges {
		if dist[edge.U] != math.MaxInt32 && dist[edge.U]+edge.Weight < dist[edge.V] {
			return nil
		}
	}

	return dist
}

func (bf *BellmanFord) HasNegativeCycle() bool {
	var start int
	for v := range bf.vertices {
		start = v
		break
	}
	return bf.ShortestPath(start) == nil
}